lib/src/rules/prefer_null_aware_method_calls.dart

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) 2021, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+
+const _desc = r'Prefer null aware method call.';
+
+const _details = r'''
+
+Instead of checking nullability of a function/method `f` before calling it you
+can use `f?.call()`.
+
+**BAD:**
+```
+if (f != null) f!();
+```
+
+**GOOD:**
+```
+f?.call();
+```
+
+''';
+
+class PreferNullAwareMethodCalls extends LintRule implements NodeLintRule {
+  PreferNullAwareMethodCalls()
+      : super(
+          name: 'prefer_null_aware_method_calls',
+          description: _desc,
+          details: _details,
+          group: Group.style,
+        );
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    var visitor = _Visitor(this);
+    registry.addIfStatement(this, visitor);
+    registry.addConditionalExpression(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitIfStatement(IfStatement node) {
+    var condition = node.condition;
+    if (condition is BinaryExpression &&
+        condition.operator.type == TokenType.BANG_EQ &&
+        condition.rightOperand is NullLiteral &&
+        node.elseKeyword == null) {
```

Comment:
Neither of these conditions check to see that the left hand side is a reference to a function, nor do they check that the function being invoked is the same as the one being tested. I think it's safe to ignore the first check as long as we check the second condition. For example, let's add a test of the form
```
var f, g;
void m() {
  if (f != null) g!();
  f != null ? g!() : null;
}
```

Reply:
The equality between condition and target of call is done at line 89. The test you mentioned is similar to lines 19 and 25 of the test file.

Reply:
Sorry, I missed that.